modules/data.land/R/IC_SOILGRID_Utilities.R

```
@@ -0,0 +1,348 @@
+  }
+  
+  # Select appropriate columns based on depth layer
+  if (depth_layer == "0-30cm") {
+    mean_c <- processed_data$data$`Total_soilC_0-30cm`[site_row]
+    original_sd <- processed_data$data$`Std_soilC_0-30cm`[site_row]
+    cv_dist <- processed_data$cv_distribution_30cm
+  } else {
+    mean_c <- processed_data$data$`Total_soilC_0-200cm`[site_row]
+    original_sd <- processed_data$data$`Std_soilC_0-200cm`[site_row]
+    cv_dist <- processed_data$cv_distribution_200cm
+  }
+  
+  if (is.na(mean_c) || mean_c <= 0) {
+    PEcAn.logger::logger.severe(sprintf("Invalid mean soil carbon value for site %s (%s)", 
+                                      site_id, depth_layer))
+  }
+  
+  soil_c_values <- numeric(size)
+  
+  # Use site-specific uncertainty
+  if (!is.na(original_sd) && original_sd > 0) {
+    shape <- (mean_c^2) / (original_sd^2)
+    rate <- mean_c / (original_sd^2)
+    if (is.finite(shape) && is.finite(rate) && shape > 0 && rate > 0) {
+      soil_c_values <- pmax(stats::rgamma(size, shape, rate), 0)
+    } else {
+      soil_c_values <- rep(mean_c, size)
+    }
+  } else if (cv_dist$type != "none") {
+    # Integrate over uncertainty using CV distribution
+    if (cv_dist$type == "gamma") {
+      cv_samples <- stats::rgamma(size, cv_dist$shape, cv_dist$rate)
+      if (all(is.finite(cv_dist$bounds))) {
+        cv_samples <- pmax(pmin(cv_samples, cv_dist$bounds[2]), cv_dist$bounds[1])
+      }
+    } else {
+      cv_samples <- sample(cv_dist$values, size, replace = TRUE)
+    }
+    
+    sd_values <- mean_c * cv_samples
+    valid <- !is.na(sd_values) & sd_values > 0
+    
+    if (any(valid)) {
+      shape_vec <- (mean_c^2) / (sd_values[valid]^2)
+      rate_vec <- mean_c / (sd_values[valid]^2)
+      soil_c_values[valid] <- pmax(stats::rgamma(sum(valid), shape_vec, rate_vec), 0)
+      soil_c_values[!valid] <- mean_c
+    } else {
+      soil_c_values <- rep(mean_c, size)
+    }
```

If the user is asking for a sampling method, signal an error if it fails rather than quietly falling back to deterministic output.

---

fixed - removed deterministic fallbacks. The function now throws explicit errors when stochastic sampling fails